### ISSUE TITLE
CA1027 false positive (Mark enums with FlagsAttribute)

### ISSUE THREAD

**Issue by patcor** [dotnet/roslyn-analyzers#6060]
### Analyzer

**Diagnostic ID**: [CA1027](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1027): `Mark enums with FlagsAttribute`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 6 SDK or later](https://docs.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 6.0.301](https://dotnet.microsoft.com/download/dotnet/6.0)

### Describe the bug

I run dotnet format below and get the following output
dotnet format .\MyProj.csproj --verify-no-changes

warning CA1027: Mark enums with FlagsAttribute [MyProj.csproj]

### Steps To Reproduce

Create a new project with the following enum and run dotnet format as described above.

```
public enum SomeData
{
	Enum0 = 0,
	Enum1 = 1,
	Enum2 = 2,
	Enum4 = 4,
	Enum6 = 6,
	Enum8 = 8,
	Enum9 = 9,
	Enum12 = 12,
	Enum13 = 13,
	Enum14 = 14,
	Enum15 = 15,
	Enum16 = 16,
	Enum17 = 17,
	Enum18 = 18,
	Enum19 = 19,
	Enum20 = 20,
	Enum21 = 21,
	Enum22 = 22,
	Enum23 = 23,
	Enum24 = 24,
	Enum25 = 25,
	Enum26 = 26,
	Enum27 = 27,
	Enum30 = 30,
	Enum31 = 31,
	Enum32 = 32,
	Enum33 = 33,
	Enum34 = 34,
	Enum35 = 35,
	Enum36 = 36,
	Enum40 = 40,
	Enum50 = 50,
	Enum51 = 51,
	Enum52 = 52,
	Enum53 = 53,
	Enum54 = 54,
	Enum55 = 55,
	Enum56 = 56,
	Enum63 = 63,
	Enum64 = 64,
	Enum65 = 65,
	Enum66 = 66,
	Enum67 = 67,
}
```

### Expected behavior
Don't expect to get a warning.

### Actual behavior
I get a warning that I should add the FlagsAttribute to a enum that clearly is not flags.

**Comment by Youssef1313** [comment]
Per the analyzer design, this seems expected.

If all non-power-of-two-values can be made of *existing* power of two values, then the analyzer produces a diagnostic.

I think a source suppression for this case is the suitable solution. @Evangelink What do you think?

**Comment by Evangelink** [comment]
Oh right! I only looked at the rule title and tried to reproduce the issue... As you said that's part of the design.

I am wondering if we could move the trigger on the potentially more False Negative (less False Positives) by looking whether the RHS. For example

```csharp
public enum MyEnum1
{
	Case1 = 0,
	Case2 = 1,
	Case3 = 2,
	Case4 = 3, // No diagnostic (or diagnostic if some option is set)
}

public enum MyEnum2
{
	Case1 = 0,
	Case2 = 1,
	Case3 = 2,
	Case4 = Case2 | Case3, // Diagnostic
}
```

WDYT?

**Comment by Youssef1313** [comment]
@Evangelink I thought of that, but I don't like introducing false negatives *especially* because this is the first and only false positive report regarding this rule.
